src/Microsoft.IdentityModel.Tokens/InMemoryCryptoProviderCache.cs

```diff
@@ -59,31 +59,39 @@ public InMemoryCryptoProviderCache() : this(new CryptoProviderCacheOptions())
-        internal InMemoryCryptoProviderCache(CryptoProviderCacheOptions cryptoProviderCacheOptions, TaskCreationOptions options, int tryTakeTimeout = 500)
+        internal InMemoryCryptoProviderCache(CryptoProviderCacheOptions cryptoProviderCacheOptions, int tryTakeTimeout = 500)
         {
-            if (cryptoProviderCacheOptions == null)
-                throw LogHelper.LogArgumentNullException(nameof(cryptoProviderCacheOptions));
-
             if (tryTakeTimeout <= 0)
                 throw LogHelper.LogArgumentException<ArgumentException>(nameof(tryTakeTimeout), $"{nameof(tryTakeTimeout)} must be greater than zero");
 
-            _cryptoProviderCacheOptions = cryptoProviderCacheOptions;
-            _signingSignatureProviders = new EventBasedLRUCache<string, SignatureProvider>(cryptoProviderCacheOptions.SizeLimit, options, StringComparer.Ordinal, false) { OnItemRemoved = (SignatureProvider signatureProvider) => signatureProvider.CryptoProviderCache = null };
-            _verifyingSignatureProviders = new EventBasedLRUCache<string, SignatureProvider>(cryptoProviderCacheOptions.SizeLimit, options, StringComparer.Ordinal, false) { OnItemRemoved = (SignatureProvider signatureProvider) => signatureProvider.CryptoProviderCache = null };
+            _cryptoProviderCacheOptions = cryptoProviderCacheOptions ?? throw LogHelper.LogArgumentNullException(nameof(cryptoProviderCacheOptions));
+            CreateProviderCache();
+        }
+
+        /// <summary>
+        /// Create the _signingSignatureProviders and _verifyingSignatureProviders caches based on the specified cache type from _cryptoProviderCacheOptions.
+        /// </summary>
+        protected void CreateProviderCache()
+        {
+            if (_cryptoProviderCacheOptions.CacheType == CryptoProviderCacheOptions.ProviderCacheType.LRU)
+            {
+                _signingSignatureProviders = new EventBasedLRUCache<string, SignatureProvider>(_cryptoProviderCacheOptions.SizeLimit, StringComparer.Ordinal) { OnItemRemoved = (SignatureProvider signatureProvider) => signatureProvider.CryptoProviderCache = null };
+                _verifyingSignatureProviders = new EventBasedLRUCache<string, SignatureProvider>(_cryptoProviderCacheOptions.SizeLimit, StringComparer.Ordinal) { OnItemRemoved = (SignatureProvider signatureProvider) => signatureProvider.CryptoProviderCache = null };
+            }
+            else
+            {
+                _signingSignatureProviders = new HashCache<string, SignatureProvider>(_cryptoProviderCacheOptions.SizeLimit, StringComparer.Ordinal) { OnItemRemoved = (SignatureProvider signatureProvider) => signatureProvider.CryptoProviderCache = null };
+                _verifyingSignatureProviders = new HashCache<string, SignatureProvider>(_cryptoProviderCacheOptions.SizeLimit, StringComparer.Ordinal) { OnItemRemoved = (SignatureProvider signatureProvider) => signatureProvider.CryptoProviderCache = null };
+            }
         }
```

Comment:
suggestion, extract to factory class?

Reply:
I was initially thinking about a factory but did not do it because this seems simple enough and may not extend further. I will make it a factory method at least.